adapter: synthesize and apply side effects (aka. controller commands) from catalog changes

aljoscha opened the pull request:
First step towards https://github.com/MaterializeInc/materialize/issues/29279

With this we only wire up the new side-effects logic for
creating/dropping tables.

### Motivation

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

def- commented:
Still in draft, but I triggered nightly: https://buildkite.com/materialize/nightly/builds/9650 and it shows an interesting panic:
```
platform-checks-mz_1-1              | thread 'coordinator' panicked at src/adapter/src/coord/ddl.rs:174:9: cannot apply ops that produce side effects when using catalog_transact_with_side_effects
```

aljoscha commented:
Would be good to get a good review from @teskje because it concerns the interaction with controllers. Maybe @petrosagg is interested as well.

And would be good for @ggevay and @SangJunBak to review as well.

@DAlperin I added you because you expressed interest.

aljoscha commented on src/adapter/src/coord/controller_commands.rs:
You're absolutely right, good catch! I initially only had `ControllerCommand`, then I added `ObjectState` to simplify the absorb method and the state machine tracking. But we don't actually need both. I'm moving `ControllerCommand` and then I'll rename `ObjectState` to `ControllerCommand`. And I'll add a description at the top of the file.

ggevay commented on src/adapter/src/coord/controller_commands/controller_state_updates.rs:
```diff
@@ -0,0 +1,285 @@
+                },
+            ) => a.cmp(b),
+            (TemporaryItem { parsed_item: a, .. }, TemporaryItem { parsed_item: b, .. }) => {
+                a.id.cmp(&b.id)
```
Unfortunately, there seems to be a different problem with this approach: CI is failing, and I think the problem is that we were relying on some ordering of catalog updates that were not enforced by timestamps, and doing consolidation on the changes to the catalog itself reordered these in an unexpected way, because consolidation calls `sort_unstable_by` under the hood. More specifically, it seems that adding builtin functions to the catalog [relies on](https://github.com/MaterializeInc/materialize/blob/53acc93eee1bbbf418fde681389aec0419db8954/src/adapter/src/catalog/builtin_table_updates.rs#L1785) having already added builtin types beforehand.

Zooming out for a moment: Can you explain why consolidation is important? What if we just let those "fluctuating" updates through? Is it just some extra work for the controllers, or would it introduce some other problem? If it's just a performance thing, then do you know how much it would affect performance?

Also, how widespread are these fluctuating updates? If it's just 1-2 things that introduce them, then maybe we could just tweak that code that is introducing them to not do that?

ggevay replied:
Ok, calling `sort_updates` after `consolidate_updates` (in the latest commit) addresses the new problem!

ggevay replied:
But could you add some comments on the `consolidate_updates` call that it is this special, tricky consolidation, where some fields are left out?

Also, could you add a comment on `transition` that it expects only at most 1 addition and at most 1 retraction for each object, and that this invariant is established by the `consolidate_updates` call in `apply_updates_inner`?

aljoscha replied:
I left out those fields in the Ord/Eq impls because those fields don't already have an Ord/Eq impl, all the in-memory catalog item stuff doesn't have them, and it would require a lot of work to add them. The assumption is that they would still cancel out, because the in-memory fields are a function of the "raw" update.

aljoscha commented:
My changes seem to have broken this branch, I'm currently going back and forth to try and find out what it is that breaks it. Please hold reviews.

teskje commented on src/adapter/src/coord/controller_commands.rs:
```diff
@@ -0,0 +1,2175 @@
+                self.retire_compute_sinks(compute_sinks_to_drop).await;
+            }
+
+            if !peeks_to_drop.is_empty() {
```
Any chance you could pull the code movement into a separate commit? I'm fine with a commit that doesn't compile, and also fine with squashing it before merging. But it would be great to remove 1000 or so lines from the diff to review, if we can :)

aljoscha commented:
@teskje and @ggevay I've split this into two parts:
- The first part adds all the new machinery, plumbing, types, etc. but doesn't move the logic over from closures in the sequencer and from the drop logic in ddl.rs.
- The second part moves over the logic. This is largely code movement, but might still benefit from having a look, to make sure things are copied over correctly. On the other hand, I think pretty much all tests would fail if DDL were not correctly applied to in-memory state and the controllers, so ... 🤷‍♂️ 

Also, an update on the discussion up ahead about adding the somewhat problematic `Ord` and `Eq` impls on the parsed state updates (see https://github.com/MaterializeInc/materialize/pull/29673#discussion_r2349298660 and up). I added a commit on top that moves the consolidation to earlier in the "pipeline", but it has some other complications that I explain in a long comment there. It's easier to look at the code of that commit and the comment than to explain here, I think.

ggevay commented:
Nightly: https://buildkite.com/materialize/nightly/builds/13501

teskje commented on src/adapter/src/coord/controller_commands/parsed_state_updates.rs:
```diff
@@ -41,7 +41,8 @@ pub enum ParsedStateUpdateKind {
         parsed_full_name: String,
     },
     TemporaryItem {
```
Is there still any need to differentiate between `ParsedStateUpdateKind::Item` and `ParsedStateUpdateKind::TemporaryItem`?

aljoscha replied:
Maybe not, but I wanted to not blur the distinction in case there is a need. The code duplication doesn't seem too horrible, but ... :man_shrugging:

teskje reviewed (commented):
Left a bunch of comments. Feel free to ignore any that complain about code that has just been moved around. I'm lacking a lot of context, so be prepared for some dumb questions as well :)

aljoscha commented on src/adapter/src/coord/controller_commands/parsed_state_updates.rs:
```diff
@@ -0,0 +1,285 @@
+) -> Option<ParsedStateUpdate> {
+    // WIP: Exhaustive match?
+    let kind = match kind {
+        StateUpdateKind::Item(item) => Some(parse_item_update(catalog, item, diff)),
+        StateUpdateKind::TemporaryItem(item) => {
+            Some(parse_temporary_item_update(catalog, item, diff))
+        }
+        StateUpdateKind::Cluster(cluster) => Some(parse_cluster_update(catalog, cluster, diff)),
+        StateUpdateKind::ClusterReplica(replica) => {
+            Some(parse_cluster_replica_update(catalog, replica, diff))
```
leftover from when the code was structured differently, I'm removing :ok_hand:

aljoscha commented on src/adapter/src/coord/controller_commands.rs:
```diff
@@ -0,0 +1,748 @@
+            &mut self,
+            item: $item_type,
+            parsed_full_name: Option<String>,
+            _ts: Timestamp,
```
will do!

aljoscha commented on src/adapter/src/coord/controller_commands.rs:
```diff
@@ -0,0 +1,748 @@
+    Sink(ControllerCommandKind<Sink>),
+    Index(ControllerCommandKind<Index>),
+    MaterializedView(ControllerCommandKind<MaterializedView>),
+    View(ControllerCommandKind<View>),
```
Ah no actually, we do need it because we drop some dependent things and clean them up when a view is dropped.

aljoscha commented on src/adapter/src/coord/controller_commands.rs:
```diff
@@ -0,0 +1,748 @@
+                    );
+                }
+                CatalogItem::ContinualTask(ct) => {
+                    self.absorb_continual_task(ct, None, catalog_update.ts, catalog_update.diff);
```
we don't use the name further on, but I'm changing it to be more consistent

aljoscha commented on src/adapter/src/coord/controller_commands.rs:
```diff
@@ -160,6 +168,1097 @@ impl Coordinator {
+                    // Nothing to do for None commands
+                }
+                ControllerCommand::Cluster(_) | ControllerCommand::ClusterReplica(_) => {
+                    unreachable!("clusters and cluster replicas are handled in a different map")
```
The map is the map of state machines that we keep, which result in the separate `commands`, `cluster_commands`, etc. vecs that we get. They are separate because they're identified by different IDs. But yeah, that code is a bit hairy.

aljoscha replied:
I could have split them up at a different level, but then we wouldn't have all the absorb business in one place.

teskje replied:
Ah, I missed the ID issue. Makes sense!

aljoscha commented on src/adapter/src/coord/controller_commands.rs:
```diff
@@ -160,6 +168,1097 @@ impl Coordinator {
+        for (catalog_id, command) in commands {
+            tracing::trace!(?command, "have to apply controller command");
+
+            match command {
```
To a degree that comes from how the code previously worked. The "add" code is from the bespoke closure that we removed. The "alter" code is from a different bespoke closure. And the drop code is from `ddl.rs`.

I think it's nice to to the batching because then you only need to do one call to, say, the storage controller, which can do things concurrently.

Re coordinator state management: yes, this is partially because it's not really only controllers. The idea is that we try and do all state changes that are required when a catalog change happens in there. So that we _can_ do these changes in adapter instances that are syncing up to catalog changes and need to update their in-memory state + controller state. That's also part of the reason why I went back-and-forth on the naming of `ParsedStateUpdate` and `ControllerStateUpdate`.

teskje replied:
Ok, I totally missed the point that this is about all adapter side effects, not just controller updates. In that case my idea of a simple dispatching `match` makes no sense and the current structure makes sense. `ParsedStateUpdate` probably also makes sense, although the name is still not ideal because its not obvious what "parsed" refers to here.

But... The reason why I assumed it was only about controller changes is that there are a lot of other things that are named to indicate that, most obviously the `controller_commands` module and the `ControllerCommand` type. Should we rename those too?

aljoscha commented on src/adapter/src/coord/controller_commands.rs:
```diff
@@ -160,6 +168,1097 @@ impl Coordinator {
+            timeline_associations.insert(timeline, (empty, id_bundle));
+        }
+
+        // No error returns are allowed after this point. Enforce this at compile time
+        // by using this odd structure so we don't accidentally add a stray `?`.
```
Yeah, you might be right. It's from here: https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/coord/ddl.rs#L605. Where it might also not be needed. But I left it as is to keep the current shape.

aljoscha commented on src/adapter/src/catalog/transact.rs:
```diff
@@ -523,9 +523,13 @@ impl Catalog {
         // We could work around this by refactoring how the interplay of
         // transact_op and apply_updates works, but that's a larger undertaking.
-        let mut preliminary_state = state.clone();
+        let mut preliminary_state = Cow::Borrowed(state);
```
will do

aljoscha commented on src/adapter/src/coord/ddl.rs:
```diff
@@ -94,9 +93,33 @@ impl Coordinator {
+
+        // WIP: Do we want to do this check here? We can't do it inside
+        // catalog_transact_inner anymore because we only apply the controller
+        // commands out here.
```
yep, I commented somewhere else that I removed these, and I removed one of the `catalog_transact_` methods

aljoscha commented:
@teskje & @ggevay I pushed a lot of commits that should address most of your comments. Could you please check:

I think there are two bigger concerns left:
- Jan commented on the naming, that it's sometimes "controller command", sometimes "parsed update". I commented inline there as well. But overall, the goal is to derive all implications from catalog changes, not just the controller commands. The title of the PR and some of the naming is therefore a bit misleading. I updated https://github.com/MaterializeInc/database-issues/issues/8488 to better describe the ultimate end goal. And that one has a parent issue that describes the outer goal, and then that one has a parent issue that describes the even broader goal ... 😅 
- I think this does slightly lower our DDL throughput, which shows up in benchmarks. We now do more work, and especially with how we derive the implications once, by updating a copy of the catalog, and then applying them again in one big batch, so that we can do the consolidation is not cheap. I think it might be fine because the ultimate goal is better scalability and I think we _can_ optimize the current process if/when we have to. @def- about those benchmarks 👈 

What do you all think?

teskje commented on src/adapter/src/catalog/apply.rs:
```diff
@@ -101,8 +102,12 @@ impl CatalogState {
-    ) -> Vec<BuiltinTableUpdate<&'static BuiltinTable>> {
+    ) -> (
+        Vec<BuiltinTableUpdate<&'static BuiltinTable>>,
+        Vec<ParsedStateUpdate>,
```
I would prefer that, yes! Should be easy to add back once we start using the updates, but until then it's nice to denoise the code a bit.

ggevay commented:
I do like `catalog_update_implications`! (But I don't count as an independent opinion, as I've lived in Germany for 6 years :D)
